### PR TITLE
Script for obtaining SQS queue messages locally

### LINKED ISSUE #452: Add a script that obtains tracebacks for all notifications in the fail queue
Automate what was done in #412.

┆Issue is synchronized with this [JIRA Story](https://ucsc-cgl.atlassian.net/browse/AZUL-303)
┆Project Name: azul
┆Issue Number: AZUL-303

### PR DESCRIPTION
**Pull request by achave11**
(fixes #452)

**NOTE:** Subject to heavy review 
* Script by default obtains all messages in queue if timeout for message
batch does not expire. This happens without deleting the messages.
* `--delete` flag will remove messages from queue.
* Required fields are `--queue-url` and `--output-file`

### CODE REVIEW

File: scripts/purge_queue.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        )
+
+        if len(response) == 0:  # Nothing left in queue
+            break
+        else:
+            for Message in response:
+                entry = {"MessageId": Message.message_id,
+                         "ReceiptHandle": Message.receipt_handle,
+                         "MD5OfBody": Message.md5_of_body,
+                         "Body": (ast.literal_eval(Message.body) if args.json_body else Message.body),
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: Why would you attempt to parse JSON with `literal_eval`? The latter parses Python literals, not serialized JSON.

File: scripts/purge_queue.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        output_file_name = args.file_name
+    Queue = sqs.Queue(url=queue_name)
+
+    seconds = 60
+    minutes = 5
+    list_of_messages = []
+    list_of_entries = []
+    while True:
+
+        response = Queue.receive_messages(
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: `messages`

File: scripts/purge_queue.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        prefix = config._index_prefix  # Limit queues displayed to our deployment prefix
+        for queue_item in list(filter(lambda queue_url: prefix in queue_url, [q.url for q in sqs.queues.all()])):
+            print(queue_item.split("/")[-1])  # Last element in array is the exact queue name!
+        sys.exit(0)
+    elif args.queue_name or args.file_name:  # If one is missing `argparse` would handle it before getting here
+        queue_name = sqs.get_queue_by_name(QueueName=args.queue_name).url
+        output_file_name = args.file_name
+    Queue = sqs.Queue(url=queue_name)
+
+    seconds = 60
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: Makes no sense.

File: scripts/purge_queue.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    minutes = 5
+    list_of_messages = []
+    list_of_entries = []
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: Compute this after you collected all messages.

File: scripts/purge_queue.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                         "Body": (ast.literal_eval(Message.body) if args.json_body else Message.body),
+                         "Attributes": Message.attributes}
+                list_of_entries.append(entry)
+            list_of_messages.append(response)
+
+    with open(output_file_name, "w") as f:
+        input = {"queue": queue_name, "messages": list_of_entries}
+        f.write(str(input))
+    manipulate_queue(Queue=Queue, messages_list=list_of_messages, delete=args.delete)
+    return sys.exit("File : '{}' written success!".format(output_file_name))
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: This signals an error.

Comment (hannes-ucsc) [reply scripts/purge_queue.py]: Also, please use a logger for all output.

File: scripts/purge_queue.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+                         "ReceiptHandle": Message.receipt_handle,
+                         "MD5OfBody": Message.md5_of_body,
+                         "Body": (ast.literal_eval(Message.body) if args.json_body else Message.body),
+                         "Attributes": Message.attributes}
+                list_of_entries.append(entry)
+            list_of_messages.append(response)
+
+    with open(output_file_name, "w") as f:
+        input = {"queue": queue_name, "messages": list_of_entries}
+        f.write(str(input))
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: Please familiarize yourself with how JSON is properly (de)serialized.

File: scripts/purge_queue.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    manipulate_queue(Queue=Queue, messages_list=list_of_messages, delete=args.delete)
+    return sys.exit("File : '{}' written success!".format(output_file_name))
+
+
+def manipulate_queue(Queue, messages_list, delete=False):
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: Don't couple the batches in which you receive messages to the batches in which you delete or return them. I can explain why in person. Collect all received messages in a simple list of messages. Then use https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.chunked to split them into batches. I recommend the use of `messages`, `message_batch` and `message_batches` for the variable / argument names.

File: scripts/purge_queue.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    for messages in messages_list:
+        if delete:
+            entries = list(map(lambda M: {"Id": M.message_id, "ReceiptHandle": M.receipt_handle}, messages))
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: `m`

A list comprehension would be more readable and shorter, I think.

File: scripts/purge_queue.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+            resp = Queue.delete_messages(Entries=entries)
+        else:
+            entries = list(map(lambda M: {"Id": M.message_id, "ReceiptHandle": M.receipt_handle,
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: See above.

File: scripts/purge_queue.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+    if args.list:
+        prefix = config._index_prefix  # Limit queues displayed to our deployment prefix
+        for queue_item in list(filter(lambda queue_url: prefix in queue_url, [q.url for q in sqs.queues.all()])):
+            print(queue_item.split("/")[-1])  # Last element in array is the exact queue name!
+        sys.exit(0)
+    elif args.queue_name or args.file_name:  # If one is missing `argparse` would handle it before getting here
+        queue_name = sqs.get_queue_by_name(QueueName=args.queue_name).url
+        output_file_name = args.file_name
+    Queue = sqs.Queue(url=queue_name)
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: This naming is nuts. `queue`

File: scripts/purge_queue.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+from azul import config
+
+parser = argparse.ArgumentParser(description=__doc__)
+
+subparsers = parser.add_subparsers(help="sub-command help")
+
+parser_dump = subparsers.add_parser("dump", help="Dump contents of queue into designated file")
+parser_dump.add_argument("queue_name", help="Name of the queue to obtain the messages from")
+parser_dump.add_argument("file_name", help="Name of file to direct output")
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
parser_dump.add_argument("file_name", help="Name of file to write output to")
```

File: scripts/purge_queue.py [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+parser_dump.add_argument("--delete", "-d", action='store_true', help="Purge messages from the queue")
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
parser_dump.add_argument("--delete", "-d", action='store_true', help="Purge messages from the queue after writing them to the designated file. By default the messages will be returned to the queue.")
```

File: scripts/purge_queue.py [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+
+parser_list = parser.add_argument("--list", "-l", help="List all available queues", action="store_true")
+
+parser_json_body = parser.add_argument("--json-body", "-jb", help="Output messages in JSON format", action="store_true")
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
parser_json_body = parser.add_argument("--json-body", "-J", help="Output messages in JSON format", action="store_true")
```

Comment (hannes-ucsc) [reply scripts/purge_queue.py]: Try to find a more precise description of what this argument does.

File: scripts/purge_queue.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+
+
+def get_messages_from_queue(args):
+    global output_file_name, queue_name
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: No globals please

File: scripts/purge_queue.py [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+    sqs = boto3.resource("sqs")
+
+    if args.list:
+        prefix = config._index_prefix  # Limit queues displayed to our deployment prefix
+        for queue_item in list(filter(lambda queue_url: prefix in queue_url, [q.url for q in sqs.queues.all()])):
+            print(queue_item.split("/")[-1])  # Last element in array is the exact queue name!
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: Use rsplit and urlparse(…).path

File: scripts/purge_queue.py [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+
+parser_json_body = parser.add_argument("--json-body", "-jb", help="Output messages in JSON format", action="store_true")
+
+
+def get_messages_from_queue(args):
+    global output_file_name, queue_name
+    sqs = boto3.resource("sqs")
+
+    if args.list:
+        prefix = config._index_prefix  # Limit queues displayed to our deployment prefix
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: _index_prefix is the prefix to ES index names.

File: scripts/purge_queue.py [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+        for queue_item in list(filter(lambda queue_url: prefix in queue_url, [q.url for q in sqs.queues.all()])):
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: `queue_item` is very misleading naming.

Comment (hannes-ucsc) [reply scripts/purge_queue.py]: `prefix in s` is ridiculous. If it's a prefix it should be `s.startswith(prefix)`

Comment (hannes-ucsc) [inline scripts/purge_queue.py]: It feels like you are now preferring `filter` and `map` over list comprehensions. A good rule of thumb is that if you have to provide a `lambda`, a comprehension will be more readable and concise. For the time being I want you to always write both forms and chose the one that more concise. You'll eventually get a feel for when to use which.

File: scripts/purge_queue.py (hunk 4 above)
Comment (achave11) [reply scripts/purge_queue.py]: @hannes-ucsc can you elaborate?

File: scripts/purge_queue.py [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+import boto3
+import json
+import more_itertools
+from azul import config
+import logging
+
+from urllib.parse import urlparse
+
+logger = logging.getLogger(__name__)
+parser = argparse.ArgumentParser(description=__doc__)
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: For this to work you need to actually provide a docstring at module scope.

File: scripts/purge_queue.py [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+
+subparsers = parser.add_subparsers(help="sub-command help", dest="command")
+
+parser_dump = subparsers.add_parser("dump", help="Dump contents of queue into designated file")
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
parser_dump = subparsers.add_parser("dump", help="Dump contents of SQS queue into specified file")
```

File: scripts/purge_queue.py [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+parser_dump.add_argument("queue_name", help="Name of the queue to obtain the messages from")
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
parser_dump.add_argument("queue_name", help="Name of the SQS queue to obtain the messages from")
```

File: scripts/purge_queue.py [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+parser_dump.add_argument("file_name", help="Name of file to write output to")
+parser_dump.add_argument("--delete", "-d", action='store_true',
+                         help="Purge messages from the queue after writing them to the designated file. "
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
                         help="Purge messages from the queue after writing them to the specified file. "
```

File: scripts/purge_queue.py [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+                              "By default the messages will be returned to the queue")
+parser_dump.add_argument("--json-body", "-J", action="store_true", help="Serialize json for queue message body")
+parser_list = subparsers.add_parser("list", help="List all available queues")
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
parser_list = subparsers.add_parser("list", help="List SQS queues in current deployment")
```

File: scripts/purge_queue.py [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+
+sqs = boto3.resource("sqs")
+
+
+def list_queues():
+    prefix = config._index_prefix  # Limit queues displayed to our ES index deployment names.
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: You can't use _index_prefix. For one, it is an internal of the Config class. More importantly, it is the prefix for Elasticsearch prefixes. Checkout indexer's app.py to see how queue names are composed.

File: scripts/purge_queue.py [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+        if queue_name.startswith(prefix):
+            print(queue_name)
+
+
+def get_messages_from_queue(args):
+    queue_name = sqs.get_queue_by_name(QueueName=args.queue_name).url
+    output_file_name = args.file_name
+    queue_resource = sqs.Queue(url=queue_name)
+    logger.info("Obtaining messages from queue '%s'", queue_name)
+    logger.info("Direct output '%s'", output_file_name)
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
    logger.info("Writing messages to '%s'", output_file_name)
```

File: scripts/purge_queue.py [hunk 25]
```diff
@@ -0,0 +1,109 @@
...
+
+    five_minutes_in_seconds = 60 * 5
+    logger.info("Default visibility timeout '%i'", output_file_name)
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: ?

File: scripts/purge_queue.py [hunk 26]
```diff
@@ -0,0 +1,109 @@
...
+    list_of_messages = []
+    while True:
+        messages = queue_resource.receive_messages(
+            AttributeNames=['All'],
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: Aligned indent please.

File: scripts/purge_queue.py [hunk 27]
```diff
@@ -0,0 +1,109 @@
...
+
+
+def get_messages_from_queue(args):
+    queue_name = sqs.get_queue_by_name(QueueName=args.queue_name).url
+    output_file_name = args.file_name
+    queue_resource = sqs.Queue(url=queue_name)
+    logger.info("Obtaining messages from queue '%s'", queue_name)
+    logger.info("Direct output '%s'", output_file_name)
+
+    five_minutes_in_seconds = 60 * 5
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
    visibility_timeout = 60 * 5
```

File: scripts/purge_queue.py [hunk 28]
```diff
@@ -0,0 +1,109 @@
...
+    logger.info("Default visibility timeout '%i'", output_file_name)
+    list_of_messages = []
+    while True:
+        messages = queue_resource.receive_messages(
+            AttributeNames=['All'],
+            MaxNumberOfMessages=10,
+            VisibilityTimeout=five_minutes_in_seconds)
+
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: Delete empty line

File: scripts/purge_queue.py [hunk 29]
```diff
@@ -0,0 +1,109 @@
...
+        if len(messages) == 0:  # Nothing left in queue
+            break
+        else:
+            list_of_messages.extend(messages)
+    with open(output_file_name, "w") as f:
+        queue_messages = json.dumps({"queue": queue_name, "messages": extract_fields(args.json_body, list_of_messages)})
+        f.write(str(queue_messages))
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: Use json.dump() to avoid buffering the JSON in memory.

File: scripts/purge_queue.py [hunk 30]
```diff
@@ -0,0 +1,109 @@
...
+    logger.info("Total messages handled: %i", len(list_of_messages))
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
    logger.info("Wrote %i messages", len(list_of_messages))
```

File: scripts/purge_queue.py [hunk 31]
```diff
@@ -0,0 +1,109 @@
...
+    message_batches = list(more_itertools.chunked(list_of_messages, 10))
+    if args.delete:
+        logger.info("Removing messages from `%s` queue", queue_name)
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
        logger.info("Removing messages from queue '%s'", queue_name)
```

Remove all backqoutes from the source.

File: scripts/purge_queue.py [hunk 32]
```diff
@@ -0,0 +1,109 @@
...
+        delete_messages(message_batches, queue_resource)
+    else:
+        logger.info("Return messages to `%s` queue", queue_name)
+        return_messages(message_batches, queue_resource)
+    print("File: {}\nWritten: OK!".format(output_file_name))
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: Use logger

File: scripts/purge_queue.py [hunk 33]
```diff
@@ -0,0 +1,109 @@
...
+
+
+def return_messages(message_batches, queue_resource):
+    for message_batch in message_batches:
+        change_visibility_batch = [dict(Id=message.message_id,
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
        batch = [dict(Id=message.message_id,
```

Comment (hannes-ucsc) [reply scripts/purge_queue.py]: The smaller the scope the shorter the variables can and should be.

File: scripts/purge_queue.py [hunk 34]
```diff
@@ -0,0 +1,109 @@
...
+                                        ReceiptHandle= message.receipt_handle,
+                                        VisibilityTimeout= 0) for message in message_batch]
+        resp = queue_resource.change_message_visibility_batch(Entries=change_visibility_batch)
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
        response = queue_resource.change_message_visibility_batch(Entries=change_visibility_batch)
```

File: scripts/purge_queue.py [hunk 35]
```diff
@@ -0,0 +1,109 @@
...
+    for queue_url in all_queues:
+        _, _, queue_name = urlparse(queue_url).path.rpartition("/")
+        if queue_name.startswith(prefix):
+            print(queue_name)
+
+
+def get_messages_from_queue(args):
+    queue_name = sqs.get_queue_by_name(QueueName=args.queue_name).url
+    output_file_name = args.file_name
+    queue_resource = sqs.Queue(url=queue_name)
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
    queue = sqs.Queue(url=queue_name)
```

Comment (hannes-ucsc) [reply scripts/purge_queue.py]: … here and everywhere else.

File: scripts/purge_queue.py [hunk 36]
```diff
@@ -0,0 +1,109 @@
...
+            raise RuntimeError(f"Failed to delete messages: entries={entries!r} resp={resp!r}")
+
+
+def extract_fields(json_body_bool, messages):
+    list_of_entries = []
+    for Message in messages:
+        entry = {"MessageId": Message.message_id,
+                 "ReceiptHandle": Message.receipt_handle,
+                 "MD5OfBody": Message.md5_of_body,
+                 "Body": (json.loads(Message.body) if json_body_bool else Message.body),
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
                 "Body": json.loads(Message.body) if json_body_bool else Message.body,
```

File: scripts/purge_queue.py [hunk 37]
```diff
@@ -0,0 +1,109 @@
...
+    for message_batch in message_batches:
+        delete_batch = [dict(Id=message.message_id, ReceiptHandle=message.receipt_handle) for message in message_batch]
+        resp = queue_resource.delete_messages(Entries=delete_batch)
+        if len(resp['Successful']) != len(delete_batch):
+            raise RuntimeError(f"Failed to delete messages: entries={entries!r} resp={resp!r}")
+
+
+def extract_fields(json_body_bool, messages):
+    list_of_entries = []
+    for Message in messages:
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
    for message in messages:
```

Comment (hannes-ucsc) [reply scripts/purge_queue.py]: -1 for missing comment from previous review.

File: scripts/purge_queue.py (hunk 18 above)
Comment (hannes-ucsc) [reply scripts/purge_queue.py]: I'd prefer not to define the parser at module scope. Put all of this and the body of the `if __name__ …` into a main() function that has the following signature: `def main(args:List[str]) -> None`. Call that function in `if __name__ …` like this `main(sys.argv[1:])`. I'll explain in person.

File: scripts/purge_queue.py [hunk 38]
```diff
@@ -0,0 +1,109 @@
...
+parser_dump = subparsers.add_parser("dump", help="Dump contents of queue into designated file")
+parser_dump.add_argument("queue_name", help="Name of the queue to obtain the messages from")
+parser_dump.add_argument("file_name", help="Name of file to write output to")
+parser_dump.add_argument("--delete", "-d", action='store_true',
+                         help="Purge messages from the queue after writing them to the designated file. "
+                              "By default the messages will be returned to the queue")
+parser_dump.add_argument("--json-body", "-J", action="store_true", help="Serialize json for queue message body")
+parser_list = subparsers.add_parser("list", help="List all available queues")
+
+sqs = boto3.resource("sqs")
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: Avoid allocating resources (any resources, not just boto3 resources) at module scope. Importing any module, even a script module should not have side-effects.

File: scripts/purge_queue.py (hunk 32 above)
Comment (achave11) [reply scripts/purge_queue.py]: 👍

**Comment by achave11** [comment]
@hannes-ucsc 
Note rejected [suggestion](https://github.com/DataBiosphere/azul/pull/492#discussion_r238113026)  Differentiation among two different types needs to be explicit. That is why there is two distinct variables.

File: scripts/purge_queue.py [hunk 39]
```diff
@@ -0,0 +1,128 @@
...
+                                                        queue_attributes["ApproximateNumberOfMessagesNotVisible"],
+                                                        queue_attributes["ApproximateNumberOfMessagesDelayed"]))
+
+
+def get_messages_from_queue(options):
+    sqs = boto3.resource("sqs")
+    logger = logging.getLogger(__name__)
+    queue_name = sqs.get_queue_by_name(QueueName=options.queue_name).url
+    output_file_name = options.file_name
+    queue = sqs.Queue(url=queue_name)
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: Why? You already had the Queue instance when you did `sqs.get_queue_by_name(QueueName=options.queue_name)` above.

File: scripts/purge_queue.py [hunk 40]
```diff
@@ -0,0 +1,128 @@
...
+    logger.info("Obtaining messages from queue '%s'", queue_name)
+    logger.info("Writing messages to '%s'", output_file_name)
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: Combine the two log messages into one.

File: scripts/purge_queue.py [hunk 41]
```diff
@@ -0,0 +1,128 @@
...
+
+    visibility_timeout = 60 * 5
+    logger.info("Default visibility timeout '%i'", visibility_timeout)
+    messages = []
+    while True:
+        messages_batch = queue.receive_messages(AttributeNames=['All'],
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: `message_batch` for the sixteenth time.

File: scripts/purge_queue.py [hunk 42]
```diff
@@ -0,0 +1,128 @@
...
+                   "messages": extract_fields(messages, options.json_body)},
+                  fp=file,
+                  indent=4)
+    logger.info("Wrote %i messages", len(messages))
+    message_batches = list(more_itertools.chunked(messages, 10))
+    if options.delete:
+        logger.info("Removing messages from queue '%s'", queue_name)
+        delete_messages(message_batches, queue)
+    else:
+        logger.info("Return messages to `%s` queue", queue_name)
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
        logger.info("Returning messages to queue `%s`", queue_name)
```

Comment (hannes-ucsc) [reply scripts/purge_queue.py]: And no backquotes in log messages. Also for the umpteenth time.

File: scripts/purge_queue.py [hunk 43]
```diff
@@ -0,0 +1,128 @@
...
+        return_messages(message_batches, queue)
+    logging.info("File: {}\nWritten: OK!".format(output_file_name))
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: I give up. Why is this using `.format()` while all the other logging code uses percents?

```suggestion
    logging.info("File: {}\nWritten: OK!".format(output_file_name))
```

File: scripts/purge_queue.py [hunk 44]
```diff
@@ -0,0 +1,128 @@
...
+            raise RuntimeError(f"Failed to revert visibility: resp={response!r}")
+
+
+def delete_messages(message_batches, queue_resource):
+    for message_batch in message_batches:
+        response = queue_resource.delete_messages(
+            Entries=[dict(Id=message.message_id,
+                          ReceiptHandle=message.receipt_handle) for message in message_batch])
+        if len(response['Successful']) != len(message_batch):
+            raise RuntimeError(f"Failed to delete messages: resp={response!r}")
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
            raise RuntimeError(f"Failed to delete messages: {response!r}")
```

Here and everywhere else.

File: scripts/purge_queue.py [hunk 45]
```diff
@@ -0,0 +1,128 @@
...
+
+
+def return_messages(message_batches, queue):
+    for message_batch in message_batches:
+        batch = [dict(Id=message.message_id,
+                      ReceiptHandle=message.receipt_handle,
+                      VisibilityTimeout=0) for message in message_batch]
+        response = queue.change_message_visibility_batch(Entries=batch)
+        if len(response['Successful']) != len(batch):
+            raise RuntimeError(f"Failed to revert visibility: resp={response!r}")
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: Why do you keep inventing new terminology? 

```suggestion
            raise RuntimeError(f"Failed to return messages: {response!r}")
```

File: scripts/purge_queue.py [hunk 46]
```diff
@@ -0,0 +1,126 @@
...
+        list_queues()
+    elif args.command == 'dump':
+        get_messages_from_queue(options=args)
+    else:
+        parser.parse_args(['--help'])
+
+
+def list_queues():
+    sqs = boto3.resource('sqs')
+    logging.info('Listing messages in queue')
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
    logging.info('Listing queues')
```

File: scripts/purge_queue.py [hunk 47]
```diff
@@ -0,0 +1,126 @@
...
+    all_queues = [q.url for q in sqs.queues.all()]
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
    all_queues = sqs.queues.all()
```

File: scripts/purge_queue.py [hunk 48]
```diff
@@ -0,0 +1,126 @@
...
+    print('\n{:<35s}{:^20s}{:^20s}{:^18s}\n'.format('Queue Name',
+                                                    'Messages Available',
+                                                    'Messages in Flight',
+                                                    'Messages Delayed'))
+    for queue_url in all_queues:
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
    for queue in all_queues:
```

File: scripts/purge_queue.py [hunk 49]
```diff
@@ -0,0 +1,126 @@
...
+        _, _, queue_name = urlparse(queue_url).path.rpartition('/')
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
        _, _, queue_name = urlparse(queue.url).path.rpartition('/')
```

File: scripts/purge_queue.py [hunk 50]
```diff
@@ -0,0 +1,126 @@
...
+        if is_azul_queue(queue_name):
+            queue = sqs.get_queue_by_name(QueueName=queue_name)
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]:
```suggestion
```

File: scripts/purge_queue.py [hunk 51]
```diff
@@ -0,0 +1,125 @@
...
+            "MessageId": message.message_id,
+            "ReceiptHandle": message.receipt_handle,
+            "MD5OfBody": message.md5_of_body,
+            "Body": json.loads(message.body) if json_body else message.body,
+            "Attributes": message.attributes
+        } for message in messages
+    ]
+
+
+def is_azul_queue(queue_name) -> bool:
```
Comment (hannes-ucsc) [inline scripts/purge_queue.py]: I think moving this here was a mistake.

Comment (achave11) [reply scripts/purge_queue.py]: @hannes-ucsc how should we handle it?

**Pull request merged**